[bug] Enable log options for Morgan that were missed in the 0.4.0 merge

mleanos opened the pull request:
Adds the log options, and format to the Morgan middleware in the Express configuration.

These options are defined in the environment configurations.

The implementation derived from https://github.com/meanjs/mean/pull/254 by @lirantal, which somehow got overlooked when merging 0.4.0 into master.

Related to the discussion in #975, and taking over for @codydaig #982

mleanos commented:
I'm aware of the coverage decrease. However, I just wanted to get this out there for review & work on writing tests in the meantime. I shall do that now :)

mleanos commented:
I wrote two tests initially for this feature. However, I'm unsure if I went about this correctly. They just seem so simple. Although, the testability of a feature like this seems limited.

It may not make sense to have the log options set in the Test env config. Does anyone have any feedback, or insight into how to enhance the test coverage here?

This brings up an interesting idea I was pondering a while ago.. Would be worthwhile to add a feature to *reinitialize* the global configuration? In instances where the NODE_ENV is manually changed (most likely for test coverage only), it would be nice to test features against the updated configuration settings. Any thought on this?

mleanos commented:
I reworked the tests, and cleaned up the default settings for the log options & format.

I think this last commit is better, but still would love some feedback.

lirantal commented:
@mleanos I did a similar "trick" with changing the NODE_ENV from testing to production through out the tests to make sure that the config works properly.

Otherwise LGTM.

mleanos commented:
@lirantal I liked your strategy of changing the NODE_ENV manually. However, that won't work in this case. Once the logger is required, the config is initialized. Changing the node env will have no impact aftward. 

What if I added a `init` method to the logger service? There I could add the `require` of the config. Is that a route worth going down?

lirantal commented:
Don't think so because that's usually a bad practice to `require` within functions/if because node can't cache that and whatever other side effects it will have.
We can go with what we have for now until we come up with another strategy.

mleanos commented:
SGTM. Thank you. 

Any opinion on if we should add the log settings to the Test env config? This way we can test the Stream option as well.

codydaig commented on config/lib/logger.js:
```diff
@@ -0,0 +1,46 @@
+      options.stream = fs.createWriteStream(process.cwd() + '/' + config.log.options.stream, { flags: 'a' });
+    } catch (err) {
+      // remove the stream option
+      delete options.stream;
```
Do we want to log an error to the console instead of just silently catching and handeling the error?

mleanos replied:
That's a good point. I think logging to the console would be appropriate here.

codydaig commented on config/lib/logger.js:
```diff
@@ -0,0 +1,46 @@
+ * Returns the log.format option set in the current environment configuration
+ */
+function getLogFormat () {
+  return config.log && config.log.format ? config.log.format : 'combined';
```
Why is this defaulting to combined? The current default is 'dev' so we want to change that?

mleanos commented:
I just committed quite a bit of changes. I added a bit more error handling, and modified some logic of how I was returning the log settings. For instance, before I was returning references to the env config settings. I think that was wrong. Now I'm making copies.

I also added a couple more tests, and added the log settings to the Test env config.

Further review would be greatly appreciated.

jloveland commented on config/env/test.js:
```diff
@@ -12,6 +12,15 @@ module.exports = {
+    // Can specify one of 'combined', 'common', 'dev', 'short', 'tiny'
+    format: process.env.LOG_FORMAT || 'combined',
+    // Stream defaults to process.stdout
+    // Uncomment to enable logging to a log on the file system
```
@mleanos what is this comment for? doesn't the option below log to the file system?

mleanos replied:
It was just a left over comment from when I copied the settings from the development config. 

I can remove it, but it also does make sense to leave it?

lirantal replied:
It seems confusing. Let's move the comments in line 18 to https://github.com/mleanos/mean/blob/morgan-logger-config/config/env/test.js#L21 inside the options object and above the stream option.

mleanos replied:
@jloveland @lirantal I moved the comment, and re-worded it. Let me know if you'd like additional changes to it. Thanks.

mleanos commented:
@codydaig I have rebased, and squashed. If there's no other feedback, or requested changes, then I think this is ready.

jloveland commented on config/lib/logger.js:
```diff
@@ -0,0 +1,80 @@
+
+      // create the WriteStream to use for the logs
+      if (options.stream.length) {
+        options.stream = fs.createWriteStream(process.cwd() + '/' + config.log.options.stream, { flags: 'a' });
```
@mleanos, great job on this!

I think it would be more robust to provide an ability to specify the full path to the directory which we would like logs to be stored. For example, I like my logs to go to `/var/log/meanjs`. Your code assumes the root directory if the cwd.

mleanos replied:
@jloveland That's a good idea. I'll work on adding this. 

I think moving the definition of the log location to be in the env config would be the way to go. This way the the `logger` config won't need to *calculate* the location of the logs.

jloveland commented:
I think it would be nice add log rotation configs to the `config.env.production.js` and have that propagate to the `config.lib.logger.js`
```js
// create a rotating write stream
var accessLogStream = FileStreamRotator.getStream({
  filename: logDirectory + '/access-%DATE%.log',
  frequency: 'daily',
  verbose: false
})

// setup the logger
app.use(morgan('combined', {stream: accessLogStream}))
```

lirantal commented:
@jloveland sounds like a good diea
@mleanos can we add the logrotating functionality?

mleanos commented:
@jloveland @lirantal The log rotating is a good idea. 

I'll have to reconfigure the stream options, but I think it can be done to easily allow switching between both log rotating, and single file log (or even possibly allow for both options).

mleanos commented:
@lirantal @jloveland I started looking at how I could change the Stream options in the env configs. Here's my initial idea... Any thoughts on this approach?

```javascript
stream: {
  single: {
    active: true,
    directoryPath: process.cwd(),
    fileName: 'access.log'
  },
  rotating: {
    active: true,
    directoryPath: 'c:/windows/logs/meanjs',
    fileName: 'access-%DATE%.log',
    frequency: 'daily',
    verbose: false
  }
}
```
We could use these settings to build the options in the `logger` config that is passed along to Morgan.

mleanos commented:
I guess the above won't work, since I can't pass both streams into Morgan. I'll keep at it.

mleanos commented:
@jloveland @lirantal I have implemented the Rotating Logs functionality. This still needs a little bit of work, and I need to write tests for the Rotating logs. But I wanted to get this out for feedback.

I had to re-work the Stream options to account for the Rotating Logs. How is this new implementation looking?

I've also refactored the tests to be more maintainable. There's one test in particular that I'm having an issue with. I have disabled this test for now, but would love to know if anyone has any insight into this. I'm getting a ENOTEMPTY error from the `fs.rmdir` call.

codydaig commented:
@mleanos I think you have to empty the directory before you can remove the directory.

mleanos commented:
@codydaig I'm using `unlink` before the `fs.rmdir` call to get rid of the file I added. Shouldn't that be sufficient?

codydaig commented:
@mleanos I would assume so as well. Sorry, I didn't get a chance to look at the code before I made that comment, just going off the error message.

lirantal commented:
Quite a lot of logic to add for the log rotation. Is there no package for that on npm that we can rely on and complicate the logic less?

mleanos commented:
@lirantal What logic are you referring to?

Don't we want to offer the option of using a single file log, or a rotating log?

lirantal commented:
Yes that's the point. I was wondering if there's some package to handle that so we don't need to start messing around with deleting files, creating directories, etc.

ilanbiala commented on package.json:
```diff
@@ -34,6 +34,7 @@
     "crypto": "0.0.3",
     "express": "^4.13.1",
     "express-session": "^1.11.3",
+    "file-stream-rotator": "0.0.6",
```
~

mleanos commented:
@lirantal I agree. I'm looking into possible packages to use. One problem is the unit testing though. If we want to write Unit Tests against the logging, then we will need to remove the test logs after the tests are complete. I'm not seeing a solution to that particular issue, unless a package offers this out of the box.

jloveland commented:
I tested these features and they LGTM!

mleanos commented:
@lirantal I ended up implementing `mock-fs` into the tests here; using this we don't have to worry about cleaning up files/directories after the tests. 

Thank you for testing with me @jloveland It's much appreciated :) 

We still have to make sure the provided log directory is created, if it doesn't exist. There's no way around it, but I don't think it will be an issue.
https://github.com/meanjs/mean/pull/999/files#diff-c31b71d5f97621da6a2b4c696f198f49R63
https://github.com/meanjs/mean/pull/999/files#diff-c31b71d5f97621da6a2b4c696f198f49R84

@codydaig @ilanbiala @lirantal @jloveland Let me know what you think of these changes, and if we're satisfied I can squash.

codydaig commented:
LGTM

ilanbiala commented on package.json:
```diff
@@ -112,6 +113,7 @@
     "karma-ng-html2js-preprocessor": "^0.1.2",
     "karma-phantomjs-launcher": "~0.2.0",
     "load-grunt-tasks": "^3.2.0",
+    "mock-fs": "^3.4.0",
```
~

ilanbiala commented:
@mleanos pretty much LGTM, just looking at tests and I'm wondering if we can add a bit more coverage for those edge cases?

https://coveralls.io/builds/3890174/source?filename=..%2Fconfig%2Flib%2Flogger.js

codydaig commented:
@ilanbiala @mleanos Something along these lines should work. I hope I remember the exact syntax correctly. 

```javascript
var spy = sinon.spy(console, 'log');

spy.firstCall.calledWith();
spy.secondCall.calledWith(chalk.yellow('Warning: An invalid format was provided. The logger will use the default format of "' + format + '"'));
spy.ThirdCall.calledWith();

console.log.restore();
```

mleanos commented:
@codydaig Thank you for the suggestion. I'll try to implement this. We have a few other areas of the app where this would be useful.

mleanos commented:
@codydaig @ilanbiala I looked into using Sinon, and it was quite handy. It turns out that using Sinon's sandbox feature, it's possible to *mute* all console logs for the duration of a test/test-suite. I think this will really be useful in a few different areas of our application.

I'd like to look at incorporating the muting of the console logs into our test coverage, but more on a global test coverage scope. Let's not worry about adding the extra coverage for the lines that log to console in this PR. I think it would be better to address it separately.

@ilanbiala I addressed the `mock-fs` dependency comment you left.

I have rebased & squashed since this should be ready now.

ilanbiala commented:
Did another once over, LGTM.

codydaig commented:
LGTM
